### PR TITLE
fix(logs): Fixing log messages for Targeted Rollouts

### PR DESCRIPTION
**Pull request by fayyazarshad**
## Summary
With this change we are introducing targeted rollout specific log messages.

## Test plan
All Unit and FSC Tests are getting passed

### CODE REVIEW

File: packages/optimizely-sdk/lib/core/bucketer/index.js [hunk 1]
```diff
@@ -112,28 +112,11 @@ export var bucket = function(bucketerParams) {
...
-    var userHasNoVariationLogMessage = sprintf(
-      LOG_MESSAGES.USER_HAS_NO_VARIATION,
-      MODULE_NAME,
-      bucketerParams.userId,
-      bucketerParams.experimentKey
-    );
-    bucketerParams.logger.log(LOG_LEVEL.DEBUG, userHasNoVariationLogMessage);
-  } else if (!bucketerParams.variationIdMap.hasOwnProperty(entityId)) {
+
+  if (!bucketerParams.variationIdMap.hasOwnProperty(entityId)) {
```
Comment (oakbani) [inline packages/optimizely-sdk/lib/core/bucketer/index.js:116]: @zashraf1985 Can you check if we really need this condition? I don't see invalid_variation handling in other SDKs.

File: packages/optimizely-sdk/lib/core/decision_service/index.js [hunk 2]
```diff
@@ -456,14 +475,14 @@ DecisionService.prototype._getVariationForRollout = function(configObj, feature,
...
+  var rolloutRule;
   var bucketerParams;
   var variationId;
   var variation;
   for (index = 0; index < endIndex; index++) {
-    experiment = configObj.experimentKeyMap[rollout.experiments[index].key];
+    rolloutRule = configObj.experimentKeyMap[rollout.experiments[index].key];
 
-    if (!this.__checkIfUserIsInAudience(configObj, experiment.key, userId, attributes)) {
+    if (!this.__checkIfUserIsInAudience(configObj, rolloutRule.key, AUDIENCE_EVALUATION_TYPES.RULE, userId, attributes)) {
```
Comment (oakbani) [inline packages/optimizely-sdk/lib/core/decision_service/index.js:485]: We need to send the log index not rollout key. I think kry for rollout is some system geenrated string. 
We would like our logs to print like ' Evaluating rollout rule 1 ... '. where 1 is an integer. Only for the last rule, we print 'EveryoneElse'

File: packages/optimizely-sdk/lib/core/bucketer/index.js (hunk 1 above)
Comment (fayyazarshad) [reply packages/optimizely-sdk/lib/core/bucketer/index.js:116]: It turns out there is a test for it as well, also it make sense to have this check as well

File: packages/optimizely-sdk/lib/core/decision_service/index.js (hunk 2 above)
Comment (fayyazarshad) [reply packages/optimizely-sdk/lib/core/decision_service/index.js]: Done

**Comment by fayyazarshad** [comment]
Resolved feedback from Owais, @zashraf1985 can you please have a look?

**Comment by fayyazarshad** [comment]
get_feature_variable_json tests of FSC are failing, due to its name is changed in https://github.com/optimizely/javascript-sdk/pull/516 which is not merged yet but testapp PR is merged with the updated name

File: packages/optimizely-sdk/lib/core/decision_service/index.js [hunk 3]
```diff
@@ -171,21 +194,24 @@ DecisionService.prototype.__getWhitelistedVariation = function(experiment, userI
...
+ * @param  {Object}  configObj            The parsed project configuration object
+ * @param  {string}  experimentKey        Key of experiment being validated
+ * @param  {string}  evaluationAttribute  String representing experiment key or rule
+ * @param  {string}  userId               ID of user
+ * @param  {Object}  attributes           Optional parameter for user's attributes
+ * @param  {string}  loggingKey           String representing experiment key or rollout rule. To be used in log messages only.
  * @return {boolean} True if user meets audience conditions
  */
-DecisionService.prototype.__checkIfUserIsInAudience = function(configObj, experimentKey, userId, attributes) {
+DecisionService.prototype.__checkIfUserIsInAudience = function(configObj, experimentKey, evaluationAttribute, userId, attributes, loggingKey) {
```
Comment (msohailhussain) [inline packages/optimizely-sdk/lib/core/decision_service/index.js:205]: Why sequence of args are changed? We don't need to rename `attributes` to evaluationAttribute.

Comment (msohailhussain) [reply packages/optimizely-sdk/lib/core/decision_service/index.js:205]: @mjc1283 In other SDKs, this method name is changed. We should rename here?

File: packages/optimizely-sdk/lib/utils/enums/index.js [hunk 4]
```diff
@@ -96,7 +96,7 @@ export var LOG_MESSAGES = {
   SHOULD_NOT_DISPATCH_ACTIVATE: '%s: Experiment %s is not in "Running" state. Not activating user.',
   SKIPPING_JSON_VALIDATION: '%s: Skipping JSON schema validation.',
   TRACK_EVENT: '%s: Tracking event %s for user %s.',
-  USER_ASSIGNED_TO_VARIATION_BUCKET: '%s: Assigned variation bucket %s to user %s.',
+  USER_ASSIGNED_TO_VARIATION_BUCKET: '%s: Assigned bucket %s to user with bucketing ID %s.',
   USER_ASSIGNED_TO_EXPERIMENT_BUCKET: '%s: Assigned experiment bucket %s to user %s.',
```
Comment (aliabbasrizvi) [inline packages/optimizely-sdk/lib/utils/enums/index.js:100]: These two should ideally be the same message. Wdyt @mjc1283 ?

File: packages/optimizely-sdk/lib/core/bucketer/index.tests.js [hunk 5]
```diff
@@ -308,7 +284,7 @@ describe('lib/core/bucketer', function() {
         it('should return null', function() {
           var bucketerParamsTest1 = cloneDeep(bucketerParams);
           bucketerParamsTest1.userId = 'ppid1';
-          expect(bucketer.bucket(bucketerParamsTest1)).to.equal('');
+          expect(bucketer.bucket(bucketerParamsTest1)).to.equal(null);
```
Comment (aliabbasrizvi) [inline packages/optimizely-sdk/lib/core/bucketer/index.tests.js:287]: How did this change?

File: packages/optimizely-sdk/lib/optimizely/index.tests.js [hunk 6]
```diff
@@ -6021,27 +6021,27 @@ describe('lib/optimizely', function() {
...
               createdLogger.log,
               LOG_LEVEL.INFO,
-              'OPTIMIZELY: Feature "test_feature" is not enabled for user user1. Returning default value for variable "new_content".'
+              'OPTIMIZELY: Feature "test_feature" is not enabled for user user1. Returning the default variable value "new_content".'
             );
             sinon.assert.calledWith(
               createdLogger.log,
               LOG_LEVEL.INFO,
-              'OPTIMIZELY: Feature "test_feature" is not enabled for user user1. Returning default value for variable "price".'
+              'OPTIMIZELY: Feature "test_feature" is not enabled for user user1. Returning the default variable value "price".'
```
Comment (aliabbasrizvi) [inline packages/optimizely-sdk/lib/optimizely/index.tests.js:6029]: The usage of `FEATURE_NOT_ENABLED_RETURN_DEFAULT_VARIABLE_VALUE` needs to be fixed in code. What I am seeing here are variable names and not variable values. Please fix and apply to this PR.

Comment (aliabbasrizvi) [reply packages/optimizely-sdk/lib/optimizely/index.tests.js:6029]: See https://github.com/optimizely/javascript-sdk/blob/8ca4242b6e3806083f6bbcdeb7e6d898a3b220e4/packages/optimizely-sdk/lib/optimizely/index.js#L820-L828. It is incorrect. Please fix the error there and adjust message to specify variable value for variable key.

File: packages/optimizely-sdk/lib/core/decision_service/index.js [hunk 7]
```diff
@@ -171,21 +194,24 @@ DecisionService.prototype.__getWhitelistedVariation = function(experiment, userI
 
 /**
  * Checks whether the user is included in experiment audience
- * @param  {Object}  configObj     The parsed project configuration object
- * @param  {string}  experimentKey Key of experiment being validated
- * @param  {string}  userId        ID of user
- * @param  {Object}  attributes    Optional parameter for user's attributes
+ * @param  {Object}  configObj            The parsed project configuration object
+ * @param  {string}  experimentKey        Key of experiment being validated
```
Comment (aliabbasrizvi) [inline packages/optimizely-sdk/lib/core/decision_service/index.js:198]: What is the purpose of `experimentKey`? Isn't `loggingKey` already fulfilling the requirement?

File: packages/optimizely-sdk/lib/core/bucketer/index.tests.js (hunk 5 above)
Comment (fayyazarshad) [reply packages/optimizely-sdk/lib/core/bucketer/index.tests.js:287]: As removed these conditions from bucketer/index.js
if (!entityId) {
    var userHasNoVariationLogMessage = sprintf(
      LOG_MESSAGES.USER_HAS_NO_VARIATION,
      MODULE_NAME,
      bucketerParams.userId,
      bucketerParams.experimentKey
    );
    bucketerParams.logger.log(LOG_LEVEL.DEBUG, userHasNoVariationLogMessage);
} else {
    var variationKey = bucketerParams.variationIdMap[entityId].key;
    var userInVariationLogMessage = sprintf(
      LOG_MESSAGES.USER_HAS_VARIATION,
      MODULE_NAME,
      bucketerParams.userId,
      variationKey,
      bucketerParams.experimentKey
    );
    bucketerParams.logger.log(LOG_LEVEL.INFO, userInVariationLogMessage);
} 
and moved these logs to decision service, previously when we had entityId = '' it use to pass the first if condition and return empty string, however it is intended to return null as per test as well. So from now on it returns null if entityId = '' or  any value doesn't exists in variationIdmap

File: packages/optimizely-sdk/lib/optimizely/index.tests.js (hunk 6 above)
Comment (fayyazarshad) [reply packages/optimizely-sdk/lib/optimizely/index.tests.js:6029]: good catch

File: packages/optimizely-sdk/lib/core/decision_service/index.js (hunk 7 above)
Comment (fayyazarshad) [reply packages/optimizely-sdk/lib/core/decision_service/index.js:198]: **experimentKey** contains experiment or rollout key, whereas **loggingKey** is being used for that index logic for **getVariationForRollout**

File: packages/optimizely-sdk/lib/utils/enums/index.js (hunk 4 above)
Comment (mjc1283) [reply packages/optimizely-sdk/lib/utils/enums/index.js:100]: Yes I agree, let's combine these @fayyazarshad.

Comment (fayyazarshad) [reply packages/optimizely-sdk/lib/utils/enums/index.js]: done

File: packages/optimizely-sdk/lib/utils/enums/index.js [hunk 8]
```diff
@@ -97,7 +97,7 @@ export var LOG_MESSAGES = {
   SKIPPING_JSON_VALIDATION: '%s: Skipping JSON schema validation.',
   TRACK_EVENT: '%s: Tracking event %s for user %s.',
   USER_ASSIGNED_TO_VARIATION_BUCKET: '%s: Assigned bucket %s to user with bucketing ID %s.',
-  USER_ASSIGNED_TO_EXPERIMENT_BUCKET: '%s: Assigned experiment bucket %s to user %s.',
+  USER_ASSIGNED_TO_EXPERIMENT_BUCKET: '%s: Assigned bucket %s to user with bucketing ID %s.',
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/utils/enums/index.js]: `USER_ASSIGNED_TO_EXPERIMENT_BUCKET` and `USER_ASSIGNED_TO_EXPERIMENT_BUCKET` are the same, so we don't need both. Please delete `USER_ASSIGNED_TO_VARIATION_BUCKET` and replace all uses of that with `USER_ASSIGNED_TO_EXPERIMENT_BUCKET`.

File: packages/optimizely-sdk/lib/utils/enums/index.js [hunk 9]
```diff
@@ -96,8 +96,8 @@ export var LOG_MESSAGES = {
   SHOULD_NOT_DISPATCH_ACTIVATE: '%s: Experiment %s is not in "Running" state. Not activating user.',
   SKIPPING_JSON_VALIDATION: '%s: Skipping JSON schema validation.',
   TRACK_EVENT: '%s: Tracking event %s for user %s.',
-  USER_ASSIGNED_TO_VARIATION_BUCKET: '%s: Assigned variation bucket %s to user %s.',
-  USER_ASSIGNED_TO_EXPERIMENT_BUCKET: '%s: Assigned experiment bucket %s to user %s.',
+  USER_ASSIGNED_TO_VARIATION_BUCKET: '%s: Assigned bucket %s to user with bucketing ID %s.',
+  USER_ASSIGNED_TO_EXPERIMENT_BUCKET: '%s: Assigned bucket %s to user with bucketing ID %s.',
```
Comment (aliabbasrizvi) [inline packages/optimizely-sdk/lib/utils/enums/index.js:100]: You can just use the same variable. Don't know why we are redeclaring.

**Comment by aliabbasrizvi** [comment]
@oakbani dismissing your review to be able to merge.

**Pull request merged**